analyzers/src/SonarAnalyzer.Common/Rules/InsteadOfAnyBase.cs

```diff
@@ -0,0 +1,124 @@
+                && Language.Syntax.TryGetOperands(invocation, out var left, out var right)
+                && IsCorrectCall(right, c.SemanticModel)
+                && c.SemanticModel.GetTypeInfo(left).Type is { } type)
+            {
+                if (ExistsTypes.Any(x => type.DerivesFrom(x)))
+                {
+                    RaiseExists(c, invocation);
+                }
+                else if (ContainsTypes.Any(x => type.DerivesFrom(x) && IsSimpleEqualityCheck(invocation, c.SemanticModel)))
+                {
+                    RaiseContains(c, invocation);
+                }
+                else if (type.DerivesFrom(KnownType.System_Collections_Generic_List_T))
+                {
+                    if (IsSimpleEqualityCheck(invocation, c.SemanticModel))
+                    {
+                        RaiseContains(c, invocation);
+                    }
+                    else
+                    {
+                        RaiseExists(c, invocation);
+                    }
+                }
+            }
+        }, Language.SyntaxKind.InvocationExpression);
+
+    protected bool IsNameEqualTo(SyntaxNode node, string name) =>
+        Language.GetName(node).Equals(name, Language.NameComparison);
+
+    protected static bool IsValueTypeOrString(SyntaxNode expression, SemanticModel model) =>
+        model.GetTypeInfo(expression).Type is { } type
+        && (type.IsValueType || type.Is(KnownType.System_String));
+
+    protected bool IsSimpleEqualsInvocation(TInvocationExpression invocation, string lambdaVariableName)
+    {
+        if (IsNameEqualTo(invocation, nameof(Equals)))
+        {
+            if (Language.Syntax.HasExactlyNArguments(invocation, 1))
+            {
+                return Language.Syntax.TryGetOperands(invocation, out var left, out _)
+                    && HasInvocationValidOperands(left, GetArgumentExpression(invocation, 0));
+            }
+            if (Language.Syntax.HasExactlyNArguments(invocation, 2))
+            {
+                return HasInvocationValidOperands(GetArgumentExpression(invocation, 0), GetArgumentExpression(invocation, 1));
+            }
+        }
+        return false;
+
+        bool HasInvocationValidOperands(SyntaxNode first, SyntaxNode second) =>
+            AreValidOperands(lambdaVariableName, first, second) || AreValidOperands(lambdaVariableName, second, first);
+    }
+
+    private static bool IsCorrectCall(SyntaxNode right, SemanticModel model) =>
+        model.GetSymbolInfo(right).Symbol is IMethodSymbol method
+        && method.IsExtensionOn(KnownType.System_Collections_Generic_IEnumerable_T);
+
+    private void RaiseExists(SonarSyntaxNodeReportingContext c, SyntaxNode invocation) =>
+        c.ReportIssue(Diagnostic.Create(existsRule, Language.Syntax.NodeIdentifier(invocation)?.GetLocation(), "Exists"));
+
+    private void RaiseContains(SonarSyntaxNodeReportingContext c, SyntaxNode invocation) =>
```

Comment:
To de-duplicate `RaiseContains` and `RaiseExists`, we can use something like this:

```csharp
    private void RaiseIssue(SonarSyntaxNodeReportingContext c, SyntaxNode invocation, DiagnosticDescriptor rule, string methodName) =>
    c.ReportIssue(Diagnostic.Create(rule, Language.Syntax.NodeIdentifier(invocation)?.GetLocation(), methodName));
```

And call is like this:

```csharp
 RaiseIssue(c, invocation, ExistsRule, "Exists");
```